### PR TITLE
Feature/testing

### PR DESCRIPTION
**Pull request by abczzz13**
## What?
Added automated testing with pytest and used Github Actions to perform these tests automaticly with every push and PR to developement. Both functional and unittests have been written to test a large part of our codebase.

## Why?
Automated tests give us a better indication to check if everything is still working properly, when making new changes in the codebase. Additionally, it gives us more confidence to refactor our code.

## How?
The tests have been written by using pytest. Additionally, pytest-django and pytest-coverage have been included for respectively a better integration with django and to check how much of our code is covered with the testing.

To test individual functions and smaller parts of our code unittests have been written. To test if the application still works as expected, functional tests have been written.

## Testing?
Tests have been included and most of the tests pass. However, a few tests have not been implemented properly and this moment and therefore fail. For now I have these tests marked as to skip, but we should fix these in future PR's.

## Anything else?
I feel like a lot tests can be written better and more efficiently. Additionally, it would be good to restructure the tests. I would suggest to do an overhaul of the tests in a future PR.

But for now, this is a good starting point to include automated testing, and tests should be included for every PR from this point. And maybe even be used as TDD?

Additionally, this PR has a lot of lines because by installing black and autopep8 all the single quotes have been updated to double quotes, which also has been included in this PR.

### CODE REVIEW

File: motivator/forms.py [hunk 1]
```diff
@@ -1,9 +1,40 @@
...
 
 
 class GoalForm(ModelForm):
-
     class Meta:
         model = Goal
-        fields = ['commit_goal', 'amount', 'start_date', 'end_date']
+        fields = ["repo", "commit_goal", "amount", "start_date", "end_date"]
+
+    # Need something as form validation
```
Comment (Reinoptland) [inline motivator/forms.py:13]: Do you mean a library? 🤔

Comment (Reinoptland) [reply motivator/forms.py:13]: I like the date validations 👍

Perhaps another validation could be wether the repo exists (easy to make a type after all), but that would require some javascript (to make a request to the github api verifying it's existence)

File: tests/motivator/functional/test_new_goal.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+@pytest.mark.parametrize(
+    "repo, commit_goal, amount, start_date, end_date, status_code",
+    [
+        (
+            "testrepo",
+            10,
+            10,
+            timezone.now(),
+            timezone.now() + timezone.timedelta(days=7),
+            302,
```
Comment (Reinoptland) [inline tests/motivator/functional/test_new_goal.py:46]: Why does this first test give a 302 status? 🤔

File: tests/motivator/functional/test_new_goal.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        ),
+        (
+            "",
```
Comment (Reinoptland) [inline tests/motivator/functional/test_new_goal.py:49]: And this one without a repo name gives 200? 😅

File: tests/motivator/functional/test_new_goal.py (hunk 2 above)
Comment (Reinoptland) [reply tests/motivator/functional/test_new_goal.py:46]: When all fields are present and the user is authenticated shouldn't it give 201 (created)

File: tests/motivator/functional/test_new_goal.py (hunk 3 above)
Comment (Reinoptland) [reply tests/motivator/functional/test_new_goal.py:49]: Shouldn't that give 400 (bad request)?

**Review by Reinoptland — COMMENTED** [review]
First of all - fantastic that you're diving into Django and immediately automatically testing you code. I like that you're combining functional tests with unit tests of more specific logic. 

Some thoughts: 

- I would like to see a validation that a certain repo or a user exist on github (maybe not now but at some point!) some validations can be synchronously at request time, some are done on a database level, but when integrating with other services you might also need validation through 3rd parties
- I like the the idea of parameterised tests as they keep the test code short. But sometimes it can be had to grasp from a list of dicts with arguments what is being tested. Here is a suggestion:

```python

@pytest.mark.parametrize(
    "username, email, github_username, status_code",
    [
        ("Test_User", "test@test.com", "testhub", 302),
        ("Test_User", "", "testhub", 200),
        ("Test_User", "test@test.com", "", 200),
    ],
)
```

VS

```python
valid_account_data  =   ("Test_User", "test@test.com", "testhub", 302),
missing_email =  ("Test_User", "", "testhub", 200),
missing_password =  ("Test_User", "test@test.com", "", 200),
@pytest.mark.parametrize(
    "username, email, github_username, status_code",
    [
      valid_account_data,
      missing_email,
      missing_password
    ],
)
```

Or something along those lines

**Comment by abczzz13** [comment]
> 

The validation of the username + repo will be implemented with the PR for retrieving the counts of commits from Github. But  yes, this is definitely something that needs to be checked.

Regarding the parameterized tests, I completely agree with you that these tests are currently hard to understand in the current implementation. I will look into this to see if it can be written in a better way.

**Comment by abczzz13** [comment]
> > 
> 
> The validation of the username + repo will be implemented with the PR for retrieving the counts of commits from Github. But yes, this is definitely something that needs to be checked.
> 
> Regarding the parameterized tests, I completely agree with you that these tests are currently hard to understand in the current implementation. I will look into this to see if it can be written in a better way.

Made an issue for both suggestions.

**Comment by Reinoptland** [comment]
@abczzz13 

The feedback on the test is sometimes explained as DRY (don't repeat yourself) vs DAMP (Descriptive And Meaningful Phrases) 

Though your tests can be DRY _and_ DAMP haha it's not either / or 

Here is a stackoverflow posts with a great answer for more context: https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests

**Comment by Reinoptland** [comment]
And making issues and improving this later is a great idea. We can get away with not validating the repo for now in the proof of concept 👍

**Pull request merged**